modules/heat_conduction/src/materials/AnisoHeatConductionMaterial.C

```diff
@@ -10,137 +10,152 @@
-  k_y = isParamValid("thermal_conductivity_y") && (NULL != _thermal_conductivity_y_pp);
-  k_z = isParamValid("thermal_conductivity_z") && (NULL != _thermal_conductivity_z_pp);
-  if (k_x || k_y || k_z)
-  {
-    mooseError("Cannot define thermal conductivity value and Postprocessor");
-  }
 }
 
+template <bool is_ad>
 void
-AnisoHeatConductionMaterial::computeProperties()
+AnisoHeatConductionMaterialTempl<is_ad>::setDerivatives(GenericRankTwoTensor<is_ad> & prop,
+                                                        RankTwoTensor dprop_dT,
+                                                        const ADReal & T)
 {
-  for (unsigned int qp(0); qp < _qrule->n_points(); ++qp)
-  {
-    (*_thermal_conductivity_x)[qp] =
-        _thermal_conductivity_x_pp ? *_thermal_conductivity_x_pp : _my_thermal_conductivity_x;
-    (*_thermal_conductivity_x_dT)[qp] = 0;
-    if (_thermal_conductivity_y)
-    {
-      (*_thermal_conductivity_y)[qp] =
-          _thermal_conductivity_y_pp ? *_thermal_conductivity_y_pp : _my_thermal_conductivity_y;
-      (*_thermal_conductivity_y_dT)[qp] = 0;
-    }
-    if (_thermal_conductivity_z)
+  unsigned int dim(_subproblem.mesh().dimension());
+  for (unsigned int i = 0; i < dim; ++i)
+    for (unsigned int j = 0; j < dim; ++j)
     {
-      (*_thermal_conductivity_z)[qp] =
-          _thermal_conductivity_z_pp ? *_thermal_conductivity_z_pp : _my_thermal_conductivity_z;
-      (*_thermal_conductivity_z_dT)[qp] = 0;
+      if (T < 0)
+        prop(i, j).derivatives() = 0.0;
+      else
+        prop(i, j).derivatives() = dprop_dT(i, j) * T.derivatives();
     }
+}
 
-    if (_specific_heat_temperature_function)
-    {
-      Point p;
-      _specific_heat[qp] = _specific_heat_temperature_function->value(_temperature[qp], p);
-    }
-    else
+template <>
+void
+AnisoHeatConductionMaterialTempl<false>::setDerivatives(GenericRankTwoTensor<false> &,
+                                                        RankTwoTensor,
+                                                        const ADReal &)
+{
+  mooseError("Mistaken call of setDerivatives in a non-AD AnisoHeatConductionMaterial");
+}
+
+template <bool is_ad>
+void
+AnisoHeatConductionMaterialTempl<is_ad>::setDerivatives(GenericReal<is_ad> & prop,
+                                                        Real dprop_dT,
+                                                        const ADReal & T)
+{
+  if (T < 0)
+    prop.derivatives() = 0;
+  else
+    prop.derivatives() = dprop_dT * T.derivatives();
+}
+
+template <>
+void
+AnisoHeatConductionMaterialTempl<false>::setDerivatives(Real &, Real, const ADReal &)
+{
+  mooseError("Mistaken call of setDerivatives in a non-AD AnisoHeatConductionMaterial version");
+}
+
+template <bool is_ad>
+void
+AnisoHeatConductionMaterialTempl<is_ad>::computeQpProperties()
+{
+  Real temp_qp = MetaPhysicL::raw_value(_T[_qp]);
+  _thermal_conductivity[_qp] = RankTwoTensor(_user_provided_thermal_conductivity);
```

Comment:
This tensor should be built once in the constructor and not over and over again at every qp

Reply:
To clarify, make `_user_provided_thermal_conductivity` a `const RankTwoTensor` (that should be all that's needed!)

Reply:
For some reason if I move the line `_thermal_conductivity[_qp] = _user_provided_thermal_conductivity;` to inside the constructor I get the following error while running the test  modules/heat_conduction/test/tests/heat_conduction_ortho/heat_conduction_ortho.i
If it is placed inside computeQpProperties() the test runs fine. 

myapp-dbgAssertion `i < _size' failed
Access out of bounds in MooseArray (i: 4294967295 size: 0)
at /Users/singh/projects/myapp/moose/framework/build/header_symlinks/MooseArray.h, line 266
Stack frames: 24
0: 0   libmesh_dbg.0.dylib                 0x00000001171a3e0c libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 460
1: 1   libmyapp-dbg.0.dylib                0x000000010036fc8d MooseArray<RankTwoTensorTempl<double> >::operator[](unsigned int) + 941
2: 2   libmyapp-dbg.0.dylib                0x0000000100372454 MaterialPropertyBase<RankTwoTensorTempl<double>, false>::operator[](unsigned int) + 36
3: 3   libmyapp-dbg.0.dylib                0x000000010047e84c AnisoHeatConductionMaterialTempl<false>::AnisoHeatConductionMaterialTempl(InputParameters const&) + 2796
4: 4   libmyapp-dbg.0.dylib                0x000000010047dd4d AnisoHeatConductionMaterialTempl<false>::AnisoHeatConductionMaterialTempl(InputParameters const&) + 29

Reply:
No, you need to initialize `_user_provided_thermal_conductivity` as a R2Tensor in the initializer list. This line her then becomes

```
_thermal_conductivity[_qp] = _user_provided_thermal_conductivity;
```

Reply:
I changed `_user_provided_thermal_conductivity` to be a R2Tensor and placed `_thermal_conductivity[_qp] = _user_provided_thermal_conductivity;` inside the `initQpStatefulProperties()`.  I get a libmesh error.  If I move `_thermal_conductivity[_qp] = _user_provided_thermal_conductivity;` to `computeQpProperties()` the error goes away, but of course that's not the right place to put it. I don't fully understand this error.

_We caught a libMesh error in ThreadedElementLoopBase
You requested to reuse the nonlinear residual vector as the base vector for computing the action of the matrix-free Jacobian, but the vectors are not the same. Your physics must have states; either remove the states from your code or make sure that you set_mf_reuse_base(false)_

See call stack in this file:
[Error.txt](https://github.com/idaholab/moose/files/6650027/Error.txt)

Reply:
Never mind, it should work now.